File: memwatch/examples/test_unified.py
#!/usr/bin/env python3
"""
memwatch - Python example with unified API
Same API and behavior as all other language bindings
"""

from memwatch import MemoryWatcher, ChangeEvent
import time

def main():
    print("MemWatch - Python Unified API Example")
    print("=" * 50)
    
    # Create watcher
    watcher = MemoryWatcher()
    print("✓ MemoryWatcher initialized")
    
    # Create a buffer to watch
    data = bytearray(b"Hello, World!")
    print(f"✓ Created buffer: {data}")
    
    # Watch it
    region_id = watcher.watch(data, name="data")
    print(f"✓ Started watching region {region_id}")
    
    # Set up callback
    events = []
    def on_change(event: ChangeEvent):
        print(f"  → Change detected: {event.variable_name}")
        print(f"    - Where: {event.where}")
        print(f"    - Old: {event.old_preview}")
        print(f"    - New: {event.new_preview}")
        events.append(event)
    
    watcher.set_callback(on_change)
    print("✓ Callback registered")
    
    # Modify data - should trigger callback
    print("\nModifying data...")
    data[0] = ord('J')  # H -> J
    time.sleep(0.1)  # Let callback process
    
    data[1] = ord('e')  # e -> e (keep it simple)
    data[2] = ord('l')
    time.sleep(0.1)
    
    # Check stats
    stats = watcher.get_stats()
    print(f"\nStats:")
    print(f"  - Tracked regions: {stats['tracked_objects']}")
    print(f"  - Total events: {len(events)}")
    
    # Cleanup
    watcher.unwatch(region_id)
    print(f"\n✓ Stopped watching region {region_id}")
    
    # Verify events were captured
    if len(events) > 0:
        print(f"\n✓ SUCCESS: Detected {len(events)} change event(s)")
        return 0
    else:
        print("\n✗ FAILURE: No change events detected")
        return 1

if __name__ == "__main__":
    exit(main())

File: memwatch/examples/test_unified.js
/**
 * memwatch - JavaScript/Node.js example with unified API
 * Same API and behavior as all other language bindings
 */

const memwatch = require('../bindings/memwatch');

async function main() {
    console.log('MemWatch - JavaScript/Node.js Unified API Example');
    console.log('='.repeat(50));
    
    // Create watcher
    const watcher = memwatch.create();
    console.log('✓ MemWatch created');
    
    // Create a buffer to watch
    const data = Buffer.from('Hello, World!');
    console.log(`✓ Created buffer: ${data.toString()}`);
    
    // Watch it
    const region_id = watcher.watch(data, 'data');
    console.log(`✓ Started watching region ${region_id}`);
    
    // Set up callback
    let events = [];
    watcher.on('change', (event) => {
        console.log(`  → Change detected: ${event.variable_name}`);
        console.log(`    - Where: ${JSON.stringify(event.where)}`);
        if (event.old_preview) console.log(`    - Old: ${event.old_preview.toString('hex').substr(0, 16)}`);
        if (event.new_preview) console.log(`    - New: ${event.new_preview.toString('hex').substr(0, 16)}`);
        events.push(event);
    });
    console.log('✓ Callback registered');
    
    // Modify data - should trigger callback
    console.log('\nModifying data...');
    data[0] = 74; // J (was H)
    await new Promise(r => setTimeout(r, 100));
    
    data.write('JavaScript', 7);
    await new Promise(r => setTimeout(r, 100));
    
    // Check stats
    const stats = watcher.get_stats();
    console.log(`\nStats:`);
    console.log(`  - Tracked regions: ${stats.num_tracked_regions}`);
    console.log(`  - Total events: ${events.length}`);
    
    // Cleanup
    watcher.unwatch(region_id);
    console.log(`\n✓ Stopped watching region ${region_id}`);
    
    // Verify events were captured
    if (events.length > 0) {
        console.log(`\n✓ SUCCESS: Detected ${events.length} change event(s)`);
        return 0;
    } else {
        console.log('\n✗ FAILURE: No change events detected');
        return 1;
    }
}

main().then(code => process.exit(code)).catch(err => {
    console.error('Error:', err);
    process.exit(1);
});
